pkg/storagev2/leveldb/store.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package leveldbstore
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	storageV2 "github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/hashicorp/go-multierror"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/errors"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+	"github.com/syndtr/goleveldb/leveldb/storage"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+type Store struct {
+	DB      *leveldb.DB
+	path    string
+	closeLk sync.RWMutex
+}
+
+var _ storageV2.Store = (*Store)(nil)
+
+// NewDatastore returns a new datastore backed by leveldb
+// for path == "", an in memory backend will be chosen (TODO)
+func New(path string, opts *opt.Options) (storageV2.Store, error) {
+	var err error
+	var db *leveldb.DB
+
+	if path == "" {
+		db, err = leveldb.Open(storage.NewMemStorage(), opts)
+	} else {
+		db, err = leveldb.OpenFile(path, opts)
+		if errors.IsCorrupted(err) && !opts.GetReadOnly() {
+			db, err = leveldb.RecoverFile(path, opts)
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	ds := Store{
+		DB:   db,
+		path: path,
+	}
+	return &ds, nil
+}
+
+func (s *Store) Count(key storageV2.Key) (c int, err error) {
+	keys := util.BytesPrefix([]byte(key.Namespace() + "/"))
+
+	iter := s.DB.NewIterator(keys, nil)
+	defer iter.Release()
+
+	for iter.Next() {
+		c++
+	}
+
+	return c, iter.Error()
+}
+
+func (s *Store) Put(item storageV2.Item) error {
+	s.closeLk.RLock()
```

Comment:
Why are we locking the db for every operation?

Reply:
We want to wait for all operations to complete before closing the database - hence the `RLock` on methods and `Lock` on the Close method.
However it looks like only the Iterate/Count operations would need this locking, because they execute multiple operations in the scope of one call, so closing the db amidst the op could lead to inconsistent behavior.
In all the other cases we could leverage the built in concurrency mechanism provided by levelDB, this is my impression after looking into library code, wdyt @aloknerurkar ?

Reply:
So for the `Close` function we need to take the Lock.

But for all the other functions we would need the RLock. This is to prevent Closing of the DB while any operations are happening. RLock can be held by multiple routines, so I dont think there would be any problem here.

Reply:
The reading of the code is strange. We are doing read-locks everywhere and then a write lock on Close for Close to wait on all locks, I get the idea but it's unconventional.

Why not go with something more readable like:

```
db.wg.Add(1)
defer db.wg.Done()
```

and then in `Close()`
```
db.wg.Wait()
```

Reply:
Sure, til the WUnlock is done, any RLocks are disallowed BUT if there are db operations waiting for the lock to be released by the Close WUnlock, then, the db operations will occur on a closed db after they acquire the lock, which may lead to a panic.

The correct implementation here is that packages relying on the localstore are closed first such that when the localstore is closed, there are no on the fly requests on the db, in which case, the wait group waiting for requests to finish may be what we need.

Reply:
> Sure, til the WUnlock is done, any RLocks are on hold BUT if there are db operations waiting for the lock to be released by the Close WUnlock, then, the db operations will occur on a closed db after they acquire the lock, which may lead to a panic.

This would not lead to panic as Close does not set the DB to nil. It should return an error I think.

> 
> The correct implementation here is that packages relying on the localstore are closed first such that when the localstore is closed, there are no on the fly requests on the db, in which case, the wait group waiting for requests to finish may be what we need. With the R and W locks, we don't have that guarantee of order of operations. 

So the case that you mentioned could happen with the Wg as well. Even after close is done an op could come in and it would increment the Wg counter and access a closed DB right? 

I mean we can never prevent user errors in using the package. If the Close is not used correctly both cases have issues in my opinion.

Reply:
> This would not lead to panic as Close does not set the DB to nil. It should return an error I think.
 Sure, don't really know.
 
> So the case that you mentioned could happen with the Wg as well. Even after close is done an op could come in and it would increment the Wg counter and access a closed DB right?

Given that we we do the rewrite correctly, and that no db operations should occurs after the close. you could have inflight requests when the db Close is called. The mutex version does not protect you like the wait group does. With the wait group, you wait for all operations to finish first at least.

Reply:
>  > This would not lead to panic as Close does not set the DB to nil. It should return an error I think.
>  
> Sure, don't really know. But the final writes on shutdown to the db could be critical, like the stamp issuers data we write on Close. So not panicking is not good enough.
>  

But so the Wg solution does not fix this. I mean this is concurrency, nothing is guaranteed. For eg when user initiates a Close it could happen that no operation was in progress so DB is closed immediately. But just after the instant Close returns, some writes are initiated. This would cause the same errors.

> > So the case that you mentioned could happen with the Wg as well. Even after close is done an op could come in and it would increment the Wg counter and access a closed DB right?
> 
> Given that we we do the rewrite correctly, and that no db operations should occurs after the close. you could have inflight requests when the db Close is called. The mutex version does not protect you like the wait group does. With the wait group, you wait for all operations to finish first at least. 
> 
> It's also more readable :)

So as I mentioned above, both solutions don't solve the problem you are referring to. If the usage is correct I would say both solutions would give you the right answer and if not both would be wrong.

In terms of readability, I am not sure. Maybe we can ask others. I would usually not use waitgroups unless I am working with go routines. Also the Close would be blocking and could cause deadlocks. So we would need to handle that again like we do all around the code right now by Waiting for the waitgroup in a different routine and then returning early if it doesn't finish in time otherwise we can hang in shutdown.